File: test_backpressure.mjs
// Runs the receiver's REAL backpressure module (not a copy of it) and asserts on
// its behaviour. Driven by test_backpressure.py so run_tests.py covers it.
//
// ⛔ WHY A NODE TEST AT ALL. The receiver is a Deno edge function and there is no
// Deno on this disk, so until now nothing in this repo could execute a single
// line of it: every "fix" to that file was verified by reading it. That is the
// exact habit standing rule 16 exists to break. backpressure.ts is deliberately
// pure - no Deno, no fetch, no globals - so node's type stripping can import and
// run it unchanged.
import {
  Breaker,
  classifyFailure,
  dropLogLine,
  httpStatusFor,
  retryDelayMs,
  BREAKER_TRIP,
  BREAKER_COOLDOWN_MS,
  RETRY_BASE_MS,
  RETRY_JITTER_MS,
} from "./supabase/functions/helius-events/backpressure.ts";

let pass = 0;
const fails = [];
function t(name, ok, detail = "") {
  if (ok) {
    pass += 1;
    console.log("  PASS  " + name + (detail ? "  [" + detail + "]" : ""));
  } else {
    fails.push(name + (detail ? "  [" + detail + "]" : ""));
    console.log("  FAIL  " + name + (detail ? "  [" + detail + "]" : ""));
  }
}

console.log("=".repeat(72));
console.log("1. the invariant the outage was caused by: NEVER 5xx on a write failure");
console.log("=".repeat(72));

for (const k of ["congestion", "privilege", "payload", "duplicate", "breaker_open"]) {
  t("⛔ httpStatusFor(" + k + ") is 200, so Helius does not redeliver",
    httpStatusFor(k) === 200, String(httpStatusFor(k)));
}

console.log("=".repeat(72));
console.log("2. classification asks 'can retrying help', not 'whose fault is it'");
console.log("=".repeat(72));

t("⛔ 403 is a PRIVILEGE failure, not congestion - the old code called it ours "
  + "and answered 500, and a 403 retried forever is pure amplification",
  classifyFailure(403) === "privilege", classifyFailure(403));
t("⛔ 401 is a privilege failure too", classifyFailure(401) === "privilege");
for (const s of [500, 502, 503, 504, 599]) {
  t("⭐ " + s + " is congestion", classifyFailure(s) === "congestion");
}
t("⭐ 429 is congestion (rate limited IS the burst problem)",
  classifyFailure(429) === "congestion");
t("⭐ 408 is congestion", classifyFailure(408) === "congestion");
for (const s of [400, 422]) {
  t("⭐ " + s + " is a payload failure - retrying a bad row can never work",
    classifyFailure(s) === "payload", classifyFailure(s));
}

// ⛔⛔ FOUND BY REPLAYING 248 REAL BATCHES AT THE LIVE RECEIVER, 2026-09-23.
// Every one came back 409 "duplicate key value violates unique constraint
// chain_events_sig_uniq", and every one was logged as chain_events_drop. The
// committed comment claimed `resolution=ignore-duplicates` made a retry a no-op.
// It did not: PostgREST emits ON CONFLICT against the PRIMARY KEY and the
// conflict is a separate unique index on `signature`. The insert now names the
// target with `?on_conflict=signature`, and this classification is the guard.
t("⛔⛔ 409 is a DUPLICATE, not a payload failure - the signature is already "
  + "stored, so nothing was lost and nothing may be reported as lost",
  classifyFailure(409) === "duplicate", classifyFailure(409));
t("⛔ and SQLSTATE 23505 in the body is a duplicate whatever the status says",
  classifyFailure(400, '{\"code\":\"23505\",\"message\":\"duplicate key\"}')
  === "duplicate");
t("⭐ a plain 400 with no 23505 is still a payload failure",
  classifyFailure(400, '{\"code\":\"22P02\"}') === "payload");
t("⛔ a 5xx that happens to mention 23505 is still CONGESTION - the database "
  + "being unwell outranks the body text",
  classifyFailure(503, "23505") === "congestion");

console.log("=".repeat(72));
console.log("3. the breaker: it OPENS, it SKIPS the database entirely, it CLOSES");
console.log("=".repeat(72));

{
  const b = new Breaker();
  const t0 = 1_000_000;
  t("⭐ a fresh breaker is closed", b.isOpen(t0) === false);
  for (let i = 1; i < BREAKER_TRIP; i++) {
    b.recordCongestion(t0);
    t("⭐ still closed after " + i + " of " + BREAKER_TRIP
      + " congestion failures - it does not trip on one unlucky request",
      b.isOpen(t0) === false);
  }
  b.recordCongestion(t0);
  t("⛔ OPEN at exactly " + BREAKER_TRIP + " consecutive congestion failures",
    b.isOpen(t0) === true);
  t("⭐ and it reports how long is left, so the answer is never a bare refusal",
    b.cooldownRemainingMs(t0) === BREAKER_COOLDOWN_MS,
    String(b.cooldownRemainingMs(t0)));
  t("⭐ still open one ms before the cooldown expires",
    b.isOpen(t0 + BREAKER_COOLDOWN_MS - 1) === true);
  t("⭐ closed the moment the cooldown is served",
    b.isOpen(t0 + BREAKER_COOLDOWN_MS) === false);
}

{
  // ⛔ The re-open case. After a served cooldown exactly one attempt goes
  // through; if the database is STILL failing, it must re-open immediately
  // rather than granting another BREAKER_TRIP free attempts.
  const b = new Breaker();
  const t0 = 2_000_000;
  for (let i = 0; i < BREAKER_TRIP; i++) b.recordCongestion(t0);
  const later = t0 + BREAKER_COOLDOWN_MS;
  t("⭐ cooldown served, so one attempt is allowed", b.isOpen(later) === false);
  b.recordCongestion(later);
  t("⛔⛔ and one further failure RE-OPENS it at once, not after another "
    + BREAKER_TRIP, b.isOpen(later) === true,
    "consecutive=" + b.consecutive);
}

{
  const b = new Breaker();
  const t0 = 3_000_000;
  for (let i = 0; i < BREAKER_TRIP; i++) b.recordCongestion(t0);
  t("⭐ open", b.isOpen(t0) === true);
  b.recordSuccess();
  t("⭐⭐ ONE successful insert closes it - the database answering is the only "
    + "evidence that matters", b.isOpen(t0) === false);
  t("⭐ and the consecutive count resets with it", b.consecutive === 0);
}

console.log("=".repeat(72));
console.log("4. a drop is LOUD and carries a signature, never a bare count");
console.log("=".repeat(72));

{
  const line = dropLogLine("congestion", 7, "5ig1abc", "503 upstream");
  const o = JSON.parse(line);
  t("⭐ the drop line is parseable JSON with its own event name",
    o.event === "chain_events_drop", o.event);
  t("⛔ it carries the COUNT", o.dropped === 7, String(o.dropped));
  t("⛔ and at least one SIGNATURE - 'some events were dropped' is not actionable",
    o.first_signature === "5ig1abc", o.first_signature);
  t("⭐ a missing signature renders as 'unknown', never as null or empty "
    + "(standing rule 5)",
    JSON.parse(dropLogLine("payload", 1, null, "x")).first_signature === "unknown");
  t("⭐ the detail is bounded so a huge PostgREST body cannot flood the log",
    JSON.parse(dropLogLine("payload", 1, "s", "y".repeat(5000)))
      .detail.length === 300);
}

console.log("=".repeat(72));
console.log("5. the retry is bounded and jittered, so isolates do not lockstep");
console.log("=".repeat(72));

t("⭐ the floor is RETRY_BASE_MS", retryDelayMs(() => 0) === RETRY_BASE_MS,
  String(retryDelayMs(() => 0)));
t("⭐ the ceiling is base + jitter",
  retryDelayMs(() => 0.9999) === RETRY_BASE_MS + RETRY_JITTER_MS - 1,
  String(retryDelayMs(() => 0.9999)));
{
  const seen = new Set();
  for (let i = 0; i < 200; i++) seen.add(retryDelayMs());
  t("⛔ and it is actually jittered, not a constant - 200 draws give many "
    + "distinct delays", seen.size > 50, "distinct=" + seen.size);
}

console.log("=".repeat(72));
console.log("6. the per-isolate counters say OUT LOUD that they are a floor");
console.log("=".repeat(72));

{
  const b = new Breaker();
  b.recordDrop("congestion", 4);
  b.recordDrop("congestion", 2);
  b.recordSkip(9);
  const snap = b.snapshot(4_000_000);
  t("⭐ drops accumulate by kind",
    snap.dropped_this_isolate_floor.congestion === 6,
    JSON.stringify(snap.dropped_this_isolate_floor));
  t("⭐ skips accumulate", snap.skipped_this_isolate_floor === 9);
  t("⛔⛔ every counter key says 'floor' IN ITS NAME, because a GET can land on "
    + "an isolate that served no POSTs and would report a false zero",
    Object.keys(snap).filter((k) => k.includes("this_isolate")).length === 2
    && Object.keys(snap).filter((k) => k.includes("this_isolate"))
      .every((k) => k.endsWith("_floor")),
    Object.keys(snap).join(","));
  t("⛔ and the snapshot carries the warning in words, not just in key names",
    /floor, never a total/.test(snap.per_isolate_warning));
}

console.log("=".repeat(72));
console.log(pass + "/" + (pass + fails.length) + " passed");
if (fails.length) {
  console.log("\nFAILED:");
  for (const f of fails) console.log("  - " + f);
  process.exit(1);
}

File: test_backpressure.py
"""The Helius receiver's backpressure, EXECUTED rather than read.

⛔⛔ WHY THIS FILE EXISTS. The receiver at `supabase/functions/helius-events/` is
a Deno edge function and there is no Deno on this disk, so until 2026-09-23 not
one line of it had ever been executed by a test. Every fix to it - eight silent
failures and counting - was verified by reading the source or by trusting a
deploy's 200. That is precisely the habit standing rule 16 exists to break, and
it cost the project's REST API at ~04:06Z that morning.

⭐ So the part that carries the load rule, `backpressure.ts`, is deliberately
PURE: no Deno, no fetch, no module globals. Node's type stripping imports and
runs it unchanged, and `test_backpressure.mjs` asserts on its behaviour. This
file runs that harness and adds the source-level invariants that can only be
checked against `index.ts`, which cannot be imported without Deno.

⚠️ A missing node is a FAILURE here, not a skip. A check that quietly does not
run is worse than one that is obviously broken.
"""
import io
import os
import re
import shutil
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
FN = os.path.join(HERE, "supabase", "functions", "helius-events")
INDEX = os.path.join(FN, "index.ts")
PURE = os.path.join(FN, "backpressure.ts")
HARNESS = os.path.join(HERE, "test_backpressure.mjs")

PASS = [0]
FAILS = []


def t(name, ok, detail=""):
    if ok:
        PASS[0] += 1
        print("  PASS  " + name + (("  [" + detail + "]") if detail else ""))
    else:
        FAILS.append(name + (("  [" + detail + "]") if detail else ""))
        print("  FAIL  " + name + (("  [" + detail + "]") if detail else ""))


def head(title):
    print("=" * 72)
    print(title)
    print("=" * 72)


src = io.open(INDEX, encoding="utf-8").read()
pure = io.open(PURE, encoding="utf-8").read()

head("1. ⛔ the receiver's pure module RUNS, in node, and passes its own suite")

node = shutil.which("node")
t("⛔⛔ node is available, so the ONLY executable test of the receiver can "
  "actually run - a silent skip here would put us straight back to verifying "
  "this file by reading it",
  bool(node), node or "NOT FOUND")

if node:
    p = subprocess.run([node, HARNESS], cwd=HERE, capture_output=True, text=True,
                       encoding="utf-8", errors="replace", timeout=120)
    out = (p.stdout or "") + (p.stderr or "")
    m = re.search(r"(\d+)/(\d+) passed", out)
    t("⭐ the harness ran to completion and reported a count",
      bool(m), (out.strip().splitlines() or ["no output"])[-1][:160])
    if m:
        t("⛔ every backpressure assertion passed",
          p.returncode == 0 and m.group(1) == m.group(2),
          m.group(0) + " rc=" + str(p.returncode))
        if p.returncode != 0:
            for line in out.splitlines():
                if line.startswith("  FAIL") or line.startswith("  - "):
                    print("      " + line.strip()[:200])

head("2. ⛔⛔ the invariant that caused the outage, checked in index.ts itself")

# The old code: `ourFault = res.status === 401 || res.status === 403 || >= 500`
# then `{ status: ourFault ? 500 : 200 }`. A 500 asks Helius to redeliver, and a
# redelivery is more load at the moment load is the problem.
t("⛔⛔ the receiver NEVER answers 5xx from the insert path - a 5xx makes "
  "Helius redeliver, and that amplification is what took out PostgREST for the "
  "whole project",
  "status: 500" not in src and "ourFault" not in src,
  "status: 500 present" if "status: 500" in src else "clean")
t("⭐ every write-failure response goes through httpStatusFor(), so the rule "
  "lives in ONE place that a test can assert on",
  src.count("httpStatusFor(") >= 3, str(src.count("httpStatusFor(")))

head("3. ⭐ the breaker is consulted BEFORE the database is touched")

# ⚠️ Anchored on the actual insert URL, not on the string "insertRows": the
# header comment now names insertRows() too, and anchoring on a name that also
# appears in prose made this assertion measure the wrong thing - it reported
# "insert at 2944", which was a position inside a COMMENT.
ins = src.index("/rest/v1/chain_events?on_conflict=signature")
t("⛔ `breaker.isOpen` is checked before the insert call appears in the file - "
  "backpressure that runs after the write is not backpressure",
  "breaker.isOpen(" in src and src.index("breaker.isOpen(") < ins,
  "isOpen at %d, insert at %d" % (src.index("breaker.isOpen("), ins))
t("⭐ and when it is open the response says how long the cooldown has left, "
  "rather than refusing without a reason",
  "cooldown_ms_remaining" in src)
t("⛔ a skipped insert is RECORDED, not silently swallowed",
  "recordSkip(" in src and "recordDrop(" in src)

head("4. ⛔ a drop is logged where a human can find it, every time")

t("⛔ every drop path writes the dropLogLine marker",
  src.count("dropLogLine(") >= 3, str(src.count("dropLogLine(")))
t("⭐ the marker has its own searchable event name",
  "chain_events_drop" in pure)
t("⛔ a PRIVILEGE failure is called out separately in the response, because it "
  "means EVERY event is being lost until a grant is fixed",
  "PRIVILEGE FAILURE" in src)

head("4b. ⛔⛔ a DUPLICATE is not a drop, and the conflict target is named")

# Found 2026-09-23 by replaying 248 real batches at the live receiver: every one
# returned 409 "duplicate key value violates unique constraint
# chain_events_sig_uniq" and every one was logged as a dropped event. The
# committed comment asserted the opposite. `Prefer: resolution=ignore-duplicates`
# does nothing on its own here, because PostgREST emits ON CONFLICT against the
# PRIMARY KEY and the conflict is a separate unique index over `signature`.
t("⛔⛔ the insert NAMES the conflict target (`on_conflict=signature`), so "
  "ignore-duplicates actually applies - without it a batch containing one "
  "already-stored signature failed atomically and took every genuinely NEW "
  "event in it down as well",
  "chain_events?on_conflict=signature" in src,
  "present" if "chain_events?on_conflict=signature" in src else "MISSING")
t("⛔ and a duplicate answers ok:true with `already_stored`, never a drop - "
  "reporting held data as lost manufactures a false alarm",
  'already_stored' in src and 'kind === "duplicate"' in src)
t("⭐ the duplicate branch records a SUCCESS on the breaker, because the "
  "database answered and is not congested",
  src.count("recordSuccess()") >= 4, str(src.count("recordSuccess()")))

head("4c. ⛔ `stored` means WRITTEN, and an unknown count is null")

t("⛔ the success path reads PostgREST's own Content-Range rather than "
  "reporting how many rows it SENT - with on_conflict working those are "
  "different numbers, and a replay reported `stored: 3` while writing nothing",
  'res.headers.get("content-range")' in src and "submitted: rows.length" in src)
t("⛔ and an unparseable header gives NULL with the reason attached, never 0 "
  "and never rows.length (standing rule 5)",
  "stored_unknown_why" in src and "inserted === null" in src)
t("⭐ count=exact is asked for, or the header would never arrive",
  "count=exact" in src)

head("5. ⛔ retrying is bounded, and only where it can possibly help")

t("⛔ only congestion is retried in process - a 401/403/400 retried has a zero "
  "success rate and is pure load",
  'if (kind === "congestion")' in src, "guard present"
  if 'if (kind === "congestion")' in src else "MISSING")
# Counted on `await insertRows()`, which is a CALL. `src.count("insertRows()")`
# also matched the declaration and a mention in a comment, so it drifted the
# moment that comment was written.
calls = src.count("await insertRows()")
t("⭐ exactly TWO calls to the insert - the first attempt and the single "
  "congestion retry, with no loop",
  calls == 2 and "while" not in src[ins - 600:ins],
  "await insertRows() appears %d times" % calls)

head("6. ⚠️ the per-isolate limit is stated, not hidden")

t("⚠️ the pure module says out loud that a per-isolate breaker is a damper "
  "and not a guarantee - the same trap the raw-payload sampler fell into, where "
  "a module-global counter could not enforce a global budget",
  "damper" in pure and "PER-ISOLATE" in pure)
t("⛔ and the self-check publishes the counters labelled as a floor",
  "backpressure: breaker.snapshot(" in src and "FLOOR" in src)

head("7. ⭐ nothing here can sign, send or move anything")

for bad in ("signTransaction", "sendTransaction", "Keypair", "secretKey",
            "setAuthority", "privateKey"):
    t("⛔ the receiver contains no `%s`" % bad, bad not in src and bad not in pure)

print("=" * 72)
print("%d/%d passed" % (PASS[0], PASS[0] + len(FAILS)))
if FAILS:
    print("\nFAILED:")
    for f in FAILS:
        print("  - " + f)
    sys.exit(1)
